src/node.ts

```diff
@@ -0,0 +1,651 @@
+import { Bytes, MarshalVersion, MetadataMapping, NodeType, Reference, StorageLoader, StorageSaver } from './types'
+import {
+  checkReference,
+  common,
+  encryptDecrypt,
+  equalBytes,
+  findIndexOfArray,
+  flattenBytesArray,
+  fromBigEndian,
+  IndexBytes,
+  keccak256Hash,
+  toBigEndianFromUint16,
+} from './utils'
+
+const pathSeparator = '/'
+
+type ForkMapping = { [key: number]: MantarayFork }
+
+const nodeForkSizes = {
+  nodeType: 1,
+  prefixLength: 1,
+  /** Bytes length before `reference` */
+  preReference: 32,
+  metadata: 2,
+  header: (): number => nodeForkSizes.nodeType + nodeForkSizes.prefixLength, // 2
+  prefixMaxSize: (): number => nodeForkSizes.preReference - nodeForkSizes.header(), // 30
+} as const
+
+const nodeHeaderSizes = {
+  obfuscationKey: 32,
+  versionHash: 31,
+  /** Its value represents how long is the `entry` in bytes */
+  refBytes: 1,
+  full: (): number => {
+    return nodeHeaderSizes.obfuscationKey + nodeHeaderSizes.versionHash + nodeHeaderSizes.refBytes
+  },
+} as const
+
+class NotFoundError extends Error {
+  constructor() {
+    super('Not found')
+  }
+}
+
+class EmptyPathError extends Error {
+  constructor() {
+    super('Empty path')
+  }
+}
+
+class UndefinedField extends Error {
+  constructor(field: string) {
+    super(`"${field}" field is not initialized.`)
+  }
+}
+
+class PropertyIsUndefined extends Error {
+  constructor() {
+    super(`Property does not exist in the object`)
+  }
+}
+
+class NotImplemented extends Error {
+  constructor() {
+    super('Not Implemented')
+  }
+}
+
+export class MantarayFork {
+  /**
+   * @param prefix the non-branching part of the subpath
+   * @param node in memory structure that represents the Node
+   */
+  constructor(public prefix: Uint8Array, public node: MantarayNode) {}
+
+  public serialize(): Uint8Array {
+    const nodeType = this.node.getType
+
+    const prefixLengthBytes = new Uint8Array(1)
+    prefixLengthBytes[0] = this.prefix.length
+
+    const prefixBytes = new Uint8Array(nodeForkSizes.prefixMaxSize())
+    prefixBytes.set(this.prefix)
+
+    const entry: Reference | undefined = this.node.getContentAddress
+
+    if (!entry) throw Error('cannot serialize MantarayFork because it does not have contentAddress')
```

Comment:
I see that there are new errors types defined, but not for every error. What is the logic behind using new error type or throwing plain `Error` objects?

Reply:
Nothing much really, that may occour more than once I make an Error class for that